File: app/src/main/java/com/example/recipeclipper/data/model/Pantry.kt
package com.example.recipeclipper.data.model

import java.util.Locale

/**
 * One thing in the pantry (#51). [name] is what the user typed (or the ingredient name of a
 * grocery line they ticked off), read with [language]'s words (null: none, so it never
 * matches). [quantity] is free text as written ("half a bag"), never read as a number.
 * [alwaysHave] marks a staple (salt, oil…): it is never on a Buy list. [purchasedDay] and
 * [expiresDay] are epoch days ([PlanDays]).
 */
data class PantryItem(
    val id: Long,
    val name: String,
    val quantity: String?,
    val language: String?,
    val aisle: Aisle,
    val inStock: Boolean,
    val alwaysHave: Boolean,
    val purchasedDay: Long?,
    val expiresDay: Long?
)

/** A new pantry item, before it has an id. It starts in stock, bought on [purchasedDay]. */
data class NewPantryItem(
    val name: String,
    val language: String?,
    val aisle: Aisle? = null,
    val quantity: String? = null,
    val purchasedDay: Long? = null
)

/** What the edit sheet can change. */
data class PantryEdit(
    val name: String,
    val quantity: String?,
    val alwaysHave: Boolean,
    val expiresDay: Long?
)

/** The expiry badge: no notifications, just a word on the row. */
enum class ExpiryBadge { EXPIRED, SOON }

enum class PantrySort { AISLE, EXPIRY }

/** The pantry as shown: one section per aisle, or (by expiry) one section with no aisle. */
data class PantrySection(val aisle: Aisle?, val items: List<PantryItem>)

/** Sorting, searching and the expiry badge. Pure. */
object PantryList {

    /** "Soon" is today and the next [SOON_DAYS] days. */
    const val SOON_DAYS = 3

    fun badge(expiresDay: Long?, today: Long): ExpiryBadge? = when {
        expiresDay == null -> null
        expiresDay < today -> ExpiryBadge.EXPIRED
        expiresDay <= today + SOON_DAYS -> ExpiryBadge.SOON
        else -> null
    }

    /**
     * [items] matching [query] (a case-insensitive part of the name; blank matches all), sorted:
     * by aisle, aisles in [Aisle] order and names A–Z within; or by expiry, soonest first, then
     * the items with no date, A–Z.
     */
    fun arrange(items: List<PantryItem>, query: String, sort: PantrySort): List<PantrySection> {
        val q = query.trim().lowercase(Locale.ROOT)
        val found = items.filter { q.isEmpty() || it.name.lowercase(Locale.ROOT).contains(q) }
        if (found.isEmpty()) return emptyList()
        val byName = compareBy<PantryItem> { it.name.lowercase(Locale.ROOT) }.thenBy { it.id }
        return when (sort) {
            PantrySort.AISLE -> found.groupBy { it.aisle }.toSortedMap(compareBy { it.ordinal })
                .map { (aisle, inAisle) -> PantrySection(aisle, inAisle.sortedWith(byName)) }
            PantrySort.EXPIRY -> listOf(
                PantrySection(
                    null,
                    found.sortedWith(compareBy<PantryItem> { it.expiresDay == null }.thenBy { it.expiresDay }.then(byName))
                )
            )
        }
    }

    /** The item already here with this name (trimmed, case-insensitive) in [language], if any. */
    fun sameName(items: List<PantryItem>, name: String, language: String?): PantryItem? {
        val key = name.trim().lowercase(Locale.ROOT)
        return items.firstOrNull { it.language == language && it.name.trim().lowercase(Locale.ROOT) == key }
    }
}

/** Have or Buy, for one ingredient of the week (#51). */
enum class NeedStatus {
    /** An in-stock pantry item has this name. Presence only: "you have flour", never "enough". */
    HAVE,

    /** A staple ("always have"): never on the Buy list, in stock or not. */
    STAPLE,

    BUY
}

/** One planned recipe's line, as the reading view renders it at the planned servings. */
data class NeedLine(
    val text: String,
    val recipeId: Long,
    val title: String,
    val day: Long?,
    val language: String?
)

/**
 * One ingredient of the week: every line naming it (exactly the same [IngredientName], in the
 * same language), and whether the pantry has it. [name] is null for a line the app can't name
 * (a heading, "salt and pepper"): it stands alone and is always Buy, since nothing is guessed.
 * [pantryName] is the matched pantry item's name.
 */
data class NeedRow(
    val name: String?,
    val lines: List<NeedLine>,
    val status: NeedStatus,
    val pantryName: String?
)

data class WeekNeeds(val buy: List<NeedRow>, val have: List<NeedRow>) {
    val isEmpty: Boolean get() = buy.isEmpty() && have.isEmpty()
}

/**
 * The pantry against a recipe's lines (#51). A line's name ([IngredientName.of]) matches a
 * pantry item's by [IngredientName.matches] ("unsalted butter" is "butter"; "rice flour" is
 * never "flour", either way), and only in the same language. It answers "is it there", never
 * "is there enough".
 */
object PantryMatch {

    /**
     * The pantry item tracking [name] in [language], or null when none is: a matching staple
     * first, else one in stock, else one that's out. [status] turns it into Have or Buy.
     */
    fun find(name: String, language: String?, pantry: List<PantryItem>): PantryItem? {
        val words = LanguageWords.forTag(language) ?: return null
        val matching = pantry.filter { it.language == words.language && IngredientName.matches(name, it.name, words) }
        return matching.firstOrNull { it.alwaysHave } ?: matching.firstOrNull { it.inStock } ?: matching.firstOrNull()
    }

    fun status(item: PantryItem?): NeedStatus = when {
        item == null -> NeedStatus.BUY
        item.alwaysHave -> NeedStatus.STAPLE
        item.inStock -> NeedStatus.HAVE
        else -> NeedStatus.BUY
    }

    /** True when [line] needn't be bought: its ingredient is in stock, or a staple. */
    fun covered(line: String, language: String?, pantry: List<PantryItem>): Boolean {
        val words = LanguageWords.forTag(language) ?: return false
        val name = IngredientName.of(line, words) ?: return false
        return status(find(name, words.language, pantry)) != NeedStatus.BUY
    }

    /**
     * The week's "What I need": every planned recipe's lines ([sources], already rendered),
     * grouped by ingredient in the order first met, split into Buy and Have. Staples are with
     * Have, never Buy.
     */
    fun weekNeeds(sources: List<GrocerySource>, pantry: List<PantryItem>): WeekNeeds {
        val groups = LinkedHashMap<Any, MutableList<NeedLine>>()
        val names = HashMap<Any, String?>()
        var unnamed = 0
        for (source in sources) {
            val words = LanguageWords.forTag(source.language)
            for (text in source.lines) {
                val line = NeedLine(text, source.recipeId, source.title, source.day, source.language)
                val name = words?.let { IngredientName.of(text, it) }
                val key: Any = if (name == null) unnamed++ else (source.language to name)
                groups.getOrPut(key) { mutableListOf() }.add(line)
                names[key] = name
            }
        }
        val rows = groups.map { (key, lines) ->
            val name = names[key]
            val item = name?.let { find(it, lines.first().language, pantry) }
            NeedRow(name, lines, status(item), item?.name)
        }
        return WeekNeeds(
            buy = rows.filter { it.status == NeedStatus.BUY },
            have = rows.filter { it.status != NeedStatus.BUY }
        )
    }
}

File: ios/RecipeClipper/Data/Model/Pantry.swift
import Foundation

/// One thing in the pantry (#51; Android's `PantryItem`). `name` is what the user typed (or the
/// ingredient name of a grocery line they ticked off), read with `language`'s words (nil: none,
/// so it never matches). `quantity` is free text as written, never read as a number.
/// `alwaysHave` marks a staple: it is never on a Buy list. The days are epoch days (`PlanDays`).
struct PantryItem: Equatable, Identifiable {
    let id: Int64
    var name: String
    var quantity: String?
    let language: String?
    var aisle: Aisle
    var inStock: Bool
    var alwaysHave: Bool
    var purchasedDay: Int64?
    var expiresDay: Int64?
}

/// A new pantry item, before it has an id. It starts in stock, bought on `purchasedDay`.
struct NewPantryItem: Equatable {
    let name: String
    let language: String?
    var aisle: Aisle? = nil
    var quantity: String? = nil
    var purchasedDay: Int64? = nil
}

/// What the edit sheet can change.
struct PantryEdit: Equatable {
    let name: String
    let quantity: String?
    let alwaysHave: Bool
    let expiresDay: Int64?
}

/// The expiry badge: no notifications, just a word on the row.
enum ExpiryBadge: Equatable { case expired, soon }

enum PantrySort: Equatable { case aisle, expiry }

/// The pantry as shown: one section per aisle, or (by expiry) one section with no aisle.
struct PantrySection: Equatable {
    let aisle: Aisle?
    let items: [PantryItem]
}

/// Sorting, searching and the expiry badge (Android's `PantryList`). Pure.
enum PantryList {

    /// "Soon" is today and the next `soonDays` days.
    static let soonDays: Int64 = 3

    static func badge(_ expiresDay: Int64?, today: Int64) -> ExpiryBadge? {
        guard let day = expiresDay else { return nil }
        if day < today { return .expired }
        if day <= today + soonDays { return .soon }
        return nil
    }

    /// `items` matching `query` (a case-insensitive part of the name; blank matches all),
    /// sorted by aisle (aisles in `Aisle` order, names A–Z within), or by expiry (soonest first,
    /// then undated, A–Z).
    static func arrange(_ items: [PantryItem], query: String, sort: PantrySort) -> [PantrySection] {
        let q = query.kTrimmed.lowercased()
        let found = items.filter { q.isEmpty || $0.name.lowercased().contains(q) }
        if found.isEmpty { return [] }
        let byName: (PantryItem, PantryItem) -> Bool = { a, b in
            let x = a.name.lowercased(), y = b.name.lowercased()
            return x != y ? x < y : a.id < b.id
        }
        switch sort {
        case .aisle:
            let order = Aisle.allCases
            return order.compactMap { aisle in
                let inAisle = found.filter { $0.aisle == aisle }
                return inAisle.isEmpty ? nil : PantrySection(aisle: aisle, items: inAisle.sorted(by: byName))
            }
        case .expiry:
            return [PantrySection(aisle: nil, items: found.sorted { a, b in
                switch (a.expiresDay, b.expiresDay) {
                case let (x?, y?): return x != y ? x < y : byName(a, b)
                case (.some, nil): return true
                case (nil, .some): return false
                case (nil, nil): return byName(a, b)
                }
            })]
        }
    }

    /// The item already here with this name (trimmed, case-insensitive) in `language`, if any.
    static func sameName(_ items: [PantryItem], name: String, language: String?) -> PantryItem? {
        let key = name.kTrimmed.lowercased()
        return items.first { $0.language == language && $0.name.kTrimmed.lowercased() == key }
    }
}

/// Have or Buy, for one ingredient of the week (#51).
enum NeedStatus: Equatable {
    /// An in-stock pantry item has this name. Presence only: "you have flour", never "enough".
    case have
    /// A staple ("always have"): never on the Buy list, in stock or not.
    case staple
    case buy
}

/// One planned recipe's line, as the reading view renders it at the planned servings.
struct NeedLine: Equatable {
    let text: String
    let recipeId: Int64
    let title: String
    let day: Int64?
    let language: String?
}

/// One ingredient of the week: every line naming it (exactly the same `IngredientName`, in the
/// same language), and whether the pantry has it. `name` is nil for a line the app can't name:
/// it stands alone and is always Buy. `pantryName` is the matched pantry item's name.
struct NeedRow: Equatable {
    let name: String?
    let lines: [NeedLine]
    let status: NeedStatus
    let pantryName: String?
}

struct WeekNeeds: Equatable {
    let buy: [NeedRow]
    let have: [NeedRow]
    var isEmpty: Bool { buy.isEmpty && have.isEmpty }
}

/// The pantry against a recipe's lines (#51; Android's `PantryMatch`). A line's name
/// (`IngredientName.of`) matches a pantry item's by `IngredientName.matches`, the density
/// table's end-of-name rule, and only in the same language. It answers "is it there", never
/// "is there enough".
enum PantryMatch {

    /// The pantry item tracking `name` in `language`, or nil: a matching staple first, else one
    /// in stock, else one that's out.
    static func find(_ name: String, language: String?, pantry: [PantryItem]) -> PantryItem? {
        guard let words = LanguageWords.forTag(language) else { return nil }
        let matching = pantry.filter { $0.language == words.language && IngredientName.matches(name, $0.name, words: words) }
        return matching.first { $0.alwaysHave } ?? matching.first { $0.inStock } ?? matching.first
    }

    static func status(_ item: PantryItem?) -> NeedStatus {
        guard let item else { return .buy }
        if item.alwaysHave { return .staple }
        return item.inStock ? .have : .buy
    }

    /// True when `line` needn't be bought: its ingredient is in stock, or a staple.
    static func covered(_ line: String, language: String?, pantry: [PantryItem]) -> Bool {
        guard let words = LanguageWords.forTag(language), let name = IngredientName.of(line, words: words) else { return false }
        return status(find(name, language: words.language, pantry: pantry)) != .buy
    }

    /// The week's "What I need": every planned recipe's lines (`sources`, already rendered),
    /// grouped by ingredient in the order first met, split into Buy and Have. Staples are with
    /// Have, never Buy.
    static func weekNeeds(_ sources: [GrocerySource], pantry: [PantryItem]) -> WeekNeeds {
        var keys: [String] = []
        var groups: [String: [NeedLine]] = [:]
        var names: [String: String] = [:]
        var unnamed = 0
        for source in sources {
            let words = LanguageWords.forTag(source.language)
            for text in source.lines {
                let line = NeedLine(text: text, recipeId: source.recipeId, title: source.title, day: source.day, language: source.language)
                let name = words.flatMap { IngredientName.of(text, words: $0) }
                let key: String
                if let name {
                    key = "n\u{1}\(source.language ?? "")\u{1}\(name)"
                    names[key] = name
                } else {
                    key = "u\u{1}\(unnamed)"
                    unnamed += 1
                }
                if groups[key] == nil { keys.append(key) }
                groups[key, default: []].append(line)
            }
        }
        let rows = keys.map { key -> NeedRow in
            let lines = groups[key]!
            let name = names[key]
            let item = name.flatMap { find($0, language: lines[0].language, pantry: pantry) }
            return NeedRow(name: name, lines: lines, status: status(item), pantryName: item?.name)
        }
        return WeekNeeds(buy: rows.filter { $0.status == .buy }, have: rows.filter { $0.status != .buy })
    }
}
